src/init_interp/initInterpMultilevelInterp.F90

```diff
@@ -546,12 +585,64 @@ subroutine interp_onelevel(data_dest, coordinate_dest, data_source, coordinates_
             errMsg(sourcefile, __LINE__))
     end if
 
+    ! Assemble the weights for scaling certain variables by soil thicknesses
+    if (scale_by_thickness) then
+       if (copylevel) then ! not using wt_lower_plus_1 and overlap_dest*
+          wt_lower = dzsoi_dest / dzsoi_source(index_lower)
+       else
+          ! Set wt_lower and overlap_dest_lower
+          SHR_ASSERT_FL((dzsoi_source(index_lower) > 0._r8), sourcefile, __LINE__)
+          overlap_dest_lower = max(0._r8, sum(dzsoi_source(1:index_lower)) - &
+                                          (dzsoi_dest_sum - dzsoi_dest))
+          wt_lower = overlap_dest_lower * dzsoi_dest / dzsoi_source(index_lower)
+
+          ! Set wt_lower_plus_one and overlap_dest_lower_plus_1
+          SHR_ASSERT_FL((dzsoi_source(index_lower+1) > 0._r8), sourcefile, __LINE__)
+          overlap_dest_lower_plus_1 = max(0._r8, &
+             dzsoi_dest_sum - sum(dzsoi_source(1:index_lower)))
+          wt_lower_plus_1 = overlap_dest_lower_plus_1 * dzsoi_dest / dzsoi_source(index_lower+1)
+          if (wt_lower <= 0._r8 .and. wt_lower_plus_1 <= 0._r8) then
+             write(iulog,*) 'Both wt_lower* are <= 0. This likely means that '
+             write(iulog,*) 'both overlap_dest* are <= 0.'
+             call endrun(msg=subname//' ERROR: This will lead to a divide by 0 in the interpolation '// &
+                  errMsg(sourcefile, __LINE__))
+          end if
+       end if
+    else  ! no scaling but still using wt_lower if (copylevel)
+       wt_lower = 1.0_r8
+    end if
+
     ! ------------------------------------------------------------------------
     ! Do the interpolation
+    ! Case 1: copylevel
+    !       Direct mapping. If scale_by_thickness=.true., we scale by
+    !       wt_lower = dzsoi_dest / dzsoi_source
+    ! Case 2a: interpolate with scale_by_thickness=.true.
+    !       The analytical solution looks like this:
+    !       data_dest = (data_source(index_lower  ) * wt_lower        + &
+    !                    data_source(index_lower+1) * wt_lower_plus_1 + &
+    !                    ...
+    !                    data_source(index_lower+n) * wt_lower_plus_n) / &
+    !                    sum(overlap_dest)
+    !       for all source layers overlapping with the destination layer. Here
+    !       we approximate with the two nearest overlapping source layers.
+    ! Case 2b: interpolate with scale_by_thickness=.false.
+    !       We use this formula:
+    !       data_dest = (data_source(index_lower  ) * distance_lower_plus_1 + &
+    !                    data_source(index_lower+1) * distance_lower) / &
+    !                   (distance_lower_plus_1 + distance_lower)
+    !       where we derive the distance variables from their respective
+    !       coordinates as an approximation of the more exact solution
+    !       derived from the corresponding overlaps between the destination
+    !       layer and each source layer.
```

Comment:
@billsacks 
originally I thought that for `scale_by_thickness=.true.` I would multiply your formula by the `dzsoi_dest / dzsoi_source` ratios for `index_lower` and `index_lower+1`; however, the new unit tests did not return expected values due to the approximation introduced when using z to calculate the distances in your formula. So I decided to split off and rewrite the solution for the `scale_by_thickness=.true.` case by using dzsoi to calculate the distances.

Reply:
@swensosc 
Bill and I would like your input and recommendation on the vertical interpolation method used by init_interp. Here's a summary of what I'm talking about:

Current method: Linear interpolation using distances between node depths. (Formula in line 647 [src/init_interp/initInterpMultilevelInterp.F90](https://github.com/ESCOMP/CTSM/pull/1301/files#diff-58791e1cb52f7439adcbdc0c14a57d56ea67c7d1bd55ce582243d8b6d62678ab) explained in line 629.)

Problem: This formula underestimates H2OSOI_LIQ and _ICE when interpolating from a high to a low resolution soil profile (and vice versa) because these variables are in kg/m2. A few kg/m2 from one or two thin layers end up in a thick layer that has room for more kg/m2, and this causes init_interp's underestimation.

Simple solution: Scale these variables and other "extensive" properties by `dzsoi_dest / dzsoi_source`. Interpolation method unchanged other than that. Destination layer will end up with double the kg/m2 coming from a source layer that's half as thick.

So far so good, until I realized that calculating the layer overlaps using node depths instead of dzsoi is (I think) less accurate. So I suggest replacing:

```
data_dest = &
(data_source(index_lower+1) * (coordinate_dest - coordinates_source(index_lower)) + &
data_source(index_lower) * (coordinates_source(index_lower+1) - coordinate_dest) &
            / (coordinates_source(index_lower+1) - coordinates_source(index_lower))
```

with...

```
data_dest = (data_source(index_lower  ) * overlap_dest_lower + &
                    data_source(index_lower+1) * overlap_dest_lower_plus_1) / &
                   (overlap_dest_lower + overlap_dest_lower_plus_1)
```

where

```
overlap_dest_lower = max(0._r8, sum(dzsoi_source(1:index_lower)) - (dzsoi_dest_sum - dzsoi_dest))
overlap_dest_lower_plus_1 = max(0._r8, dzsoi_dest_sum - sum(dzsoi_source(1:index_lower)))
```

I have a [google doc](https://docs.google.com/presentation/d/1ka1QkXqZcekQqCmj5S1eMGaEgM-pxZy5jwvvnNoasj4/edit?usp=sharing) that may help visualize what I'm talking about. Let me know if it would be more efficient to have a 30-minute meeting to go over this stuff. Either way Bill and I are interested in your feedback. Thanks!